### ISSUE TITLE
Can't change/set passwords with SASL auth

### ISSUE THREAD

**Issue by SunilMohanAdapa** [martymac/ldapscripts#5]
Currently, there is no way to set or change the password while using SASL authentication.

In FreedomBox, a very simplified UI manages everything including LDAP user accounts.  The administrative interface has to take care of managing the user accounts.  Since storing LDAP admin password on the system somewhere is not good, we are using SASL Auth EXTERNAL and connecting via ldapi:/// URL to manage the users.  We have modified the permissions as necessary.

With this approach, we are unable to change a user's password or set it during user creation as `ldapscripts` refuses to do so.  The relevant code looks as follows:

``` sh
    if [ -n "$SASLAUTH" ]
    then
      # XXX Is there a reason to allow changing a userPassword attribute here ?
      end_die "Please, change password in $SASLAUTH database"
```

I believe there is a have realistic use case and we should allow changing the password for the user using `ldappasswd` command even when using SASL authentication.

**Comment by martymac** [comment]
On Mon, 27 Jul 2015 11:01:01 -0700, Sunil Mohan Adapa wrote

Hi Sunil,

> I believe there is a have realistic use case and we should 
> allow changing the password for the user using `ldappasswd` 
> command even when using SASL authentication.

Thanks for this report and use case :)

Does the following patch work for you ?

```
--- lib/runtime.orig
+++ lib/runtime
@@ -692,11 +692,7 @@ _changepassword () {
       echo "$2 : $1" >> "$PASSWORDFILE"
     fi

-    if [ -n "$SASLAUTH" ]
-    then
-      # XXX Is there a reason to allow changing a userPassword attribute here ?
-      end_die "Please, change password in $SASLAUTH database"
-    elif [ -n "$BINDPWDFILE" ]
+    if [ -n "$SASLAUTH" ] || [ -n "$BINDPWDFILE" ]
     then
       ## Change password in a secure way
       # Allocate and create temp file
@@ -704,7 +700,12 @@ _changepassword () {
       # Generate password file
       echo -n "$1" > "$_TMPFILE" || end_die "Error writing to temporary file $_TMPFILE"
       # Change password
-      $LDAPPASSWDBIN $LDAPBINOPTS -y "$BINDPWDFILE" -D "$BINDDN" -xH "$SERVER" -T "$_TMPFILE" "$2" 2>>"$LOGFILE" 1>/dev/null
+      if [ -n "$SASLAUTH" ]
+      then
+        $LDAPPASSWDBIN $LDAPBINOPTS -Y "$SASLAUTH" -H "$SERVER" -T "$_TMPFILE" "$2" 2>>"$LOGFILE" 1>/dev/null
+      else # [ -n "$BINDPWDFILE" ]
+        $LDAPPASSWDBIN $LDAPBINOPTS -y "$BINDPWDFILE" -D "$BINDDN" -xH "$SERVER" -T "$_TMPFILE" "$2" 2>>"$LOGFILE" 1>/dev/null
+      fi
       _RESULT=$?
       # Release temp file
       reltempf
```

Best regards,

## 

Ganael LAPLANCHE ganael.laplanche@martymac.org
http://www.martymac.org | http://contribs.martymac.org
FreeBSD: martymac martymac@FreeBSD.org, http://www.FreeBSD.org

**Comment by SunilMohanAdapa** [comment]
On 08/07/2015 01:31 AM, Ganael Laplanche wrote:

> On Mon, 27 Jul 2015 11:01:01 -0700, Sunil Mohan Adapa wrote
> 
> Hi Sunil,
> 
> > I believe there is a have realistic use case and we should
> > allow changing the password for the user using `ldappasswd`
> > command even when using SASL authentication.
> 
> Thanks for this report and use case :)
> 
> Does the following patch work for you ?

The patch works as expected for my SASLAUTH EXTERNAL case.

Thank you very much for the fix.  Sorry about the late response.

It would be nice if you can tag a new release with the fixes.  I can
request for the fixes to be packaged in Debian (and consequently in
FreedomBox).

## 

Sunil

**Comment by martymac** [comment]
Committed, thanks Sunil!

I'll try to release 2.0.6 soon :)

Best regards,

Ganael.

**Comment by martymac** [comment]
Hi Sunil,

I've just released ldapscripts-2.0.6 :)

Best regards,
Ganael.

**Comment by SunilMohanAdapa** [comment]
Awesome. Thank you again. I have requested the Debian package maintainer to release a newer version as well.
